### PR TITLE
client: periodically reload InClusterConfig token

### PR DESCRIPTION
**Pull request by mikedanese**
/sig auth
/sig api-machinery

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/client-go/rest/in_cluster_config_test.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+	for i, c := range tests {
+		t.Run(fmt.Sprint(i), func(t *testing.T) {
+			tts := &testTokenSource{
+				tok: c.tsTok,
+				err: c.tsErr,
+			}
+
+			ts := &cachingTokenSource{
+				base:   tts,
+				tok:    c.tok,
```
[the remark this answers is not in the document]
Comment (mikedanese) [reply staging/src/k8s.io/client-go/rest/in_cluster_config_test.go]: I have test cases that do that.

File: staging/src/k8s.io/client-go/rest/in_cluster_config.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+	return tst.base.RoundTrip(req)
+}
+
+type fileTokenSource struct {
+	path   string
+	period time.Duration
+}
+
+func (ts *fileTokenSource) Token() (*oauth2.Token, error) {
+	tokb, err := ioutil.ReadFile(ts.path)
```
[the remark this answers is not in the document]
Comment (liggitt) [reply staging/src/k8s.io/client-go/rest/in_cluster_config.go]: Disagree here. We don't want to change the behavior of how we handle token files. That could lead to token files with leading/trailing spaces that work with this (common) client library and fail with others that expect the data to be good

**Comment by liggitt** [comment]
limiting this to InClusterConfig seems like the safest first step. Are there equivalent methods in the client libraries for other languages we need to update?

Would you expect the same updating behavior to apply to referenced token files eventually? I know of pods that run components using a kubeconfig file constructed with caFile/tokenFile references that point to the injected serviceaccount ca/token paths.

**Comment by mikedanese** [comment]
I would expect this to work tor token-file as well. I'd like to replace the BearerToken field in rest.Config with a oauth2.TokenSource to unify all these methods. I can do that now or later.

File: staging/src/k8s.io/client-go/rest/token_source.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+	}
+
+	// slow path
+	ts.Lock()
+	defer ts.Unlock()
+	if tok := ts.tok; tok != nil && tok.Expiry.Add(-1*ts.leeway).After(now) {
+		return tok, nil
+	}
+
+	tok, err := ts.base.Token()
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/rest/token_source.go]: if we previously successfully read a token, and failed to re-read a token, I think we should continue to use the previously read token.

File: staging/src/k8s.io/client-go/rest/token_source.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+	}
+}
+
+func newCachedPathTokenSource(path string) oauth2.TokenSource {
+	return &cachingTokenSource{
+		now:    time.Now,
+		leeway: 1 * time.Minute,
+		base: &fileTokenSource{
+			path:   path,
+			period: 5 * time.Minute,
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/rest/token_source.go]: comment explaining why this period was chosen would be good

**Comment by liggitt** [comment]
a couple comments, looks reasonable to me otherwise. 

if token reloading is desired/required in the future (at least for in-cluster configs), this should be noted in the [kubernetes client guidelines](https://docs.google.com/document/d/1hsJNlowIg-u_rz3JBw9hXh6rj2LgeAdMY7OleOL1srw/edit?disco=AAAABIUzD4o), and issues should be opened to track adding it to the client libraries in other languages

File: staging/src/k8s.io/client-go/rest/token_source.go (hunk 4 above)
Comment (mikedanese) [reply staging/src/k8s.io/client-go/rest/token_source.go]: If only there was a reason...

File: staging/src/k8s.io/client-go/rest/token_source.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+func newCachedPathTokenSource(path string) oauth2.TokenSource {
+	return &cachingTokenSource{
+		now:    time.Now,
+		leeway: 1 * time.Minute,
+		base: &fileTokenSource{
+			path: path,
+			// This period was picked because it is half of the minimum validity
+			// duration for a token provisioned by they TokenRequest API. This is
+			// unsophisticated and should induce rotation that a frequency that
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/rest/token_source.go]: "rotation *at* a frequency"?

File: staging/src/k8s.io/client-go/rest/token_source.go [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+	// slow path
+	ts.Lock()
+	defer ts.Unlock()
+	if tok := ts.tok; tok != nil && tok.Expiry.Add(-1*ts.leeway).After(now) {
+		return tok, nil
+	}
+
+	tok, err := ts.base.Token()
+	if err != nil {
+		if ts.tok == nil || ts.tok.Expiry.After(now) {
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/rest/token_source.go]: I don't get the expiry check here... using the last read token seems better than stopping using it based on an arbitrary expiry

**Comment by liggitt** [comment]
one nit, one question about error handling, lgtm otherwise

File: staging/src/k8s.io/client-go/rest/token_source.go (hunk 6 above)
Comment (mikedanese) [reply staging/src/k8s.io/client-go/rest/token_source.go]: How should we surface the errors?

Comment (liggitt) [reply staging/src/k8s.io/client-go/rest/token_source.go]: newCachedPathTokenSource can be told whether to log them, I suppose. looks like the call site already logs some errors using glog.Errorf

File: staging/src/k8s.io/client-go/rest/token_source.go (hunk 5 above)
Comment (mikedanese) [reply staging/src/k8s.io/client-go/rest/token_source.go]: Done

File: staging/src/k8s.io/client-go/rest/token_source.go (hunk 6 above)
Comment (mikedanese) [reply staging/src/k8s.io/client-go/rest/token_source.go]: Done

**Comment by mikedanese** [comment]
Godeps approval:

/assign @cblecker

**Comment by cblecker** [comment]
No root dep changes, just new packages in staging repos. As long as the -verify job passes, I'm cool with this.
/approve

**Comment by cblecker** [comment]
`staging/src/k8s.io/sample-cli-plugin/Godeps/OWNERS` huh.. when did this come in. doesn't look like it's using the staging godeps script.

/hold
Holding until I can check this..

**Comment by cblecker** [comment]
Oh, it is.. but we don't enforce the owners file like we do for root. Okay, this is fine until I can follow up with the OWNERS changes.

/hold cancel

**Comment by mikedanese** [comment]
@cblecker I sent one out a few minutes ago:

https://github.com/kubernetes/kubernetes/pull/68154

**Comment by mikedanese** [comment]
@liggitt Done.

**Comment by liggitt** [comment]
Can you also open an issue to track https://github.com/kubernetes/kubernetes/pull/67359#issuecomment-415463989 so we don't forget to fold this back into the guidelines and other clients?

**Comment by cblecker** [comment]
/lgtm cancel
Failing tests

**Comment by cblecker** [comment]
The last directory is the csi-api Godeps, that was added after the fact. I'm going to submit a PR to fix, but I'm also going to manually approve to get this in

**Comment by timothysc** [comment]
@kubernetes/sig-auth-bugs @kubernetes/sig-api-machinery-bugs - Folks - we're still digging into the source, but so far we believe this is the prime suspect which caused a regression in the in-cluster configuration that is used e2e test framework.

**Comment by timothysc** [comment]
Submitted - https://github.com/kubernetes/kubernetes/pull/69272

**Pull request merged**